segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorageTest.java

```diff
@@ -0,0 +1,296 @@
+
+        int offset = 0;
+
+        try ( Storage s1 = createStorage()) {
+            s1.initialize(DEFAULT_EPOCH);
+
+            s1.create(segmentName, TIMEOUT).join();
+            s1.create(concatSegmentName, TIMEOUT).join();
+
+            SegmentHandle writeHandle1 = s1.openWrite(segmentName).join();
+            SegmentHandle writeHandle2 = s1.openWrite(concatSegmentName).join();
+
+            byte[] writeData = String.format("Segment_%s_Append", segmentName).getBytes();
+            ByteArrayInputStream dataStream1 = new ByteArrayInputStream(writeData);
+            ByteArrayInputStream dataStream2 = new ByteArrayInputStream(writeData);
+
+            s1.write(writeHandle1, offset, dataStream1, writeData.length, TIMEOUT).join();
+            s1.write(writeHandle2, offset, dataStream2, writeData.length, TIMEOUT).join();
+
+            s1.seal(writeHandle2, TIMEOUT).join();
+
+            // This will append the segments and delete the concat segment.
+            s1.concat(writeHandle1, writeData.length, concatSegmentName, TIMEOUT).join();
+            long lengthBeforeRetry = s1.getStreamSegmentInfo(segmentName, TIMEOUT).join().getLength();
+
+            // Create the segment again.
+            s1.create(newConcatSegmentName, TIMEOUT).join();
+            writeHandle2 = s1.openWrite(newConcatSegmentName).join();
+            dataStream2 = new ByteArrayInputStream(writeData);
+            s1.write(writeHandle2, offset, dataStream2, writeData.length, TIMEOUT).join();
+            s1.seal(writeHandle2, TIMEOUT).join();
+
+            //Concat at the same offset again
+            s1.concat(writeHandle1, writeData.length, newConcatSegmentName, TIMEOUT).join();
+            long lengthAfterRetry = s1.getStreamSegmentInfo(segmentName, TIMEOUT).join().getLength();
+            Assert.assertTrue( String.format("Concatenation of same segment at the same offset(%d) should result in " +
+                            "same segment size(%d), but is (%d)", writeData.length, lengthBeforeRetry,
+                    lengthAfterRetry),
+                    lengthBeforeRetry == lengthAfterRetry);
+
+            //Verify the data
+            byte[] readBuffer = new byte[writeData.length];
+            for (int j = 0; j < 2; j++) {
+                int bytesRead = s1.read(writeHandle1, j * readBuffer.length, readBuffer,
+                        0, readBuffer.length, TIMEOUT) .join();
+                Assert.assertEquals(String.format("Unexpected number of bytes read from offset %d.", offset),
+                        readBuffer.length, bytesRead);
+                AssertExtensions.assertArrayEquals(String.format("Unexpected read result from offset %d.", offset),
+                        readBuffer, (int) offset, readBuffer, 0, bytesRead);
+            }
+            s1.delete(writeHandle1, TIMEOUT).join();
+        }
+    }
+
+    //endregion
+
+    //Region RetryHelper tests
+    @Test
+    public void retry() throws Exception {
```

Comment:
Please DO NOT put this unit test here. It doesn't test code in `FileSystemStorage`. Put it in `FileSystemRetryHelperTest` - or if you choose to use the standard retry framework as I suggested above, this may not be necessary.